[master] Fix 64433: Add dynamic loading of file_roots, pillar_roots, and thorium_roots

bluesliverx opened the pull request:
### What does this PR do?
Adds dynamic expansion of file/pillar/thorium roots config.

### What issues does this PR fix or reference?
Fixes: #64433

### Previous Behavior
The roots were expanded only once at startup.

### New Behavior
The roots are expanded on every access of the environments within the `file_roots`, `pillar_roots`, and `thorium_roots` options.

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Commit:
Add dynamic loading of file_roots, pillar_roots, and thorium_roots

Ch3LL commented on salt/config/__init__.py:
```diff
@@ -1753,55 +1754,43 @@ def _gather_buffer_space():
     """
-    for saltenv, dirs in file_roots.items():
-        normalized_saltenv = str(saltenv)
-        if normalized_saltenv != saltenv:
```
I'm not seeing this if logic replicated anywhere. Am I possibly missing it?

bluesliverx replied:
It is covered by line 1775, since the previous implementation iterated over the file roots and would pop and insert entries to it. The new version just creates an entirely new dynamic dictionary and therefore doesn't need to do this comparison and/or pop any entries in the roots.

Ch3LL reviewed (changes requested):
Can we get some additional test coverage that actually tests the full functionality. For example, adding a new directory path to file roots and being able to call that SLS file and same for pillar.

I'd like to also get some additional reviews on this.

Commit:
Add integration tests for dynamic configuration and improve docs

bluesliverx commented:
@Ch3LL I have added tests and updated the documentation as requested. Let me know if you see anything else.

bluesliverx commented:
Friendly ping for @dwoz/@twangboy for reviews :)

dwoz commented:
Looked at this. I like the feature but am not sure about DynamicDict just yet. I'm going to chew on it for a bit before giving a final review.

bluesliverx commented:
> Looked at this. I like the feature but am not sure about DynamicDict just yet. I'm going to chew on it for a bit before giving a final review.

Thanks @dwoz, I should give you some background that the DynamicDict was written by someone else at my company, and it fell to me to get it into upstream Salt. The first thing I did is sat back and consider if there was a better way to do this, such as using a method in `salt.config` to retrieve the file roots or updating the opts when changes were detected (too hard for config on disparate filesystems such as NFS). In the end I ended up back with the dynamic dict since it provides a seamless interface to module writers, etc. i.e. `opts['file_roots']` will work in either case with no changes to salt modules or third party modules.

We have been using pretty much this exact implementation for about 5 years now on clusters of 30k+ of salt minions, and it works well for us. With that said, I'm open to suggestions, just let me know.

bluesliverx commented:
@dwoz, not trying to be pushy here, just curious if you had any further thoughts on this? (just in case it dropped off your radar :) )

dwoz commented:
Sorry about the delayed response here. I was hoping to come around to allowing the DynamicDict thing to fly. However, that hasn't happened. I do no't like the 'magic' of dynamic dict and I do not like adding something other than simple data types to config and subsequently opts. If we're going to implement this feature, and I think we should. I think it'd be better to have a method in config or utils to grab the file_roots as needed and where needed.

bluesliverx commented:
@dwoz I appreciate the honest feedback, and completely understand. Should the method in config/utils handle updating the file_roots or be used for actual retrieval? My only worry about doing it there is that something (even an execution module not in this repo) will use file_roots that will not be updated with the latest changes.

dwoz commented:
@bluesliverx Looks like this is causing some tests to fail

bluesliverx commented:
@dwoz, I'm a bit confused. You said you didn't want this PR merged in its current form. I'm looking for some feedback on what exactly you were thinking (see my last comment for questions there) before I rework it and bring it up to date with the latest master. Or has that opinion changed?

dwoz commented:
> @dwoz, I'm a bit confused. You said you didn't want this PR merged in its current form. I'm looking for some feedback on what exactly you were thinking (see my last comment for questions there) before I rework it and bring it up to date with the latest master. Or has that opinion changed?

I don't have the bandwidth to dive in and try and come up with a better solution. If the tests all passed I'd be more inclined to let it go without being my preferred implementation. 

@s0undt3ch Thoughts on this one?

bluesliverx commented:
Sorry @twangboy, thanks for the review! I'm still fixing up some tests that were affected by this change. I'm getting close but since I can't seem to run tests locally anymore (not sure why it freezes for me now but worked previously...), so I just have to run it up here unfortunately.

twangboy commented:
No worries. Let me know when you're ready for a review.

bluesliverx commented:
@twangboy I *think* this might be ready, but it's no longer running the build for some reason, so hard to tell. I was really close on the last commit though.

bluesliverx commented:
@twangboy @dwoz this should be ready now.

twangboy commented:
Sometimes it just takes a while for the jobs to start

twangboy commented:
Not sure why those tests are failing

bluesliverx commented:
@twangboy, yeah, me either. I've been trying to watch master, because I believe it is failing to build master in general and has nothing to do with my change: https://github.com/saltstack/salt/actions?query=branch%3Amaster

twangboy commented:
Yeah, we've made a lot of fixes on 3006.x. Those are now being merged forward to 3007.x and then to master. Maybe the fixes are there.

twangboy commented:
This will need a review from @dwoz

twangboy commented:
You got some pre-commit and lint failures that need to be addressed

Commit:
Fix formatting

twangboy commented on tests/pytests/unit/utils/test_dynamic_dict.py:
```diff
@@ -0,0 +1,193 @@
+def test_delete_dyn_key(ddict):
+    ddict.add_dyn("foo", _dyn_func)
+    val = ddict.pop("foo")
+    assert val == _dyn_func(), f"Failed pop(): {val} != {dyn_func()}"
```
Looks like this is causing the Lint failure:
```
tests/pytests/unit/utils/test_dynamic_dict.py:188: [E0602(undefined-variable), test_delete_dyn_key] Undefined variable 'dyn_func'
```

bluesliverx replied:
Thanks! Forgot to check back on this and suddenly I can no longer run linting locally with pre-commit, still gotta figure that out... but this should be fixed.

Commit:
Cleanup salt minion key after test

Commit:
Update test to copy config first

twangboy commented:
@bluesliverx Would you mind rebasing and resolving the conflicts?

bluesliverx commented:
@twangboy, sure thing. Apologies that this has been languishing. I have one consistently failed test that I can't seem to get to pass, and without a windows machine I'm finding this extremely difficult to test locally.

twangboy commented on tests/pytests/integration/cli/test_salt_key.py:
```diff
@@ -172,6 +172,8 @@ def test_list_all_no_check_files(
     with pytest.helpers.change_cwd(str(config_dir)):
         master_config = copy.deepcopy(salt_master.config)
+        # Convert the file roots to a static dict before appending any roots
+        master_config["file_roots"] = master_config["file_roots"].static_dict()
```
Maybe try skipping this line on Windows?
```
if not salt.utils.platform.is_windows():
```

bluesliverx replied:
I added that to attempt to fix it as that was the only thing I could see affecting it. It didn't make a difference either way though.

Commit:
Fix merge conflicts

Commit:
Fix testing for yaml dumper

Commit:
Update version to specify v3008 for dynamic file roots

twangboy commented on doc/ref/configuration/master.rst:
```diff
@@ -4264,6 +4277,19 @@ is equivalent to this static configuration:
       prod:
         - /srv/prod/pillar
 
+As of 3007.0, the ``pillar_roots`` option is dynamically expanded on each use,
```
3008.0

bluesliverx replied:
Doh, updated one reference but missed this one.

bluesliverx commented:
@twangboy, this is ready now I believe.

bluesliverx commented:
How does one go about getting this merged? :)

twangboy commented:
I'm thinking we'll get things fixed on 3006.x and 3007.x, then merge those forward into master. Then we can rebase and get this in.

twangboy commented:
When we get closer to release we will start merging pending PRs into master branch that are passing tests.

dwoz commented:
@bluesliverx needs rebase

bluesliverx commented:
@dwoz rebased

twangboy commented:
Please address the merge conflicts

bluesliverx commented:
@twangboy done

twangboy commented on salt/utils/dynamic_dict.py:
```diff
@@ -0,0 +1,92 @@
+    """
+    A dictionary that can mix static and dynamic values.
+    """
+
+    def __init__(self, *args, **argv):
+        self._func_dict = {}
+        super().__init__(*args, **argv)
+
+    def __getitem__(self, key):
+        val = super().__getitem__(key)
+        if key in self._func_dict:
+            val = self._func_dict[key](val, dyn_dict=self, key=key)
+        return val
```
Critical Performance Bug: Uncached Disk I/O on Dictionary Access
The Issue: Every time a core Salt subsystem reads an environment path (e.g., checking ``opts["file_roots"]["base"])``, ``__getitem__`` triggers the ``_expand_glob_path`` callback. This function performs expensive, blocking file system I/O using ``glob.glob(path)``.

The Impact: During a highstate run or compilation, the options dictionary is accessed hundreds or thousands of times. Running uncached disk globbing loops repeatedly on every single key lookup will severely degrade master performance at scale.

The Fix: Implement a short-lived cache (TTL mechanism) inside ``DynamicDict`` so that the expensive glob evaluation hits the physical disk at a reasonable interval rather than on every single key read.

A clean way to solve this is to add an internal cache dictionary inside ``DynamicDict`` that tracks a timestamp alongside the evaluated paths, limiting disk scans to once every few seconds.

Modify the ``DynamicDict`` class in ``salt/utils/dynamic_dict.py``:

```python
import time
import glob

class DynamicDict(dict):
    def __init__(self, *args, **argv):
        self._func_dict = {}
        self._cache = {}
        self._ttl = 5.0  # Cache disk globs for 5 seconds
        super().__init__(*args, **argv)

    def __getitem__(self, key):
        val = super().__getitem__(key)
        if key in self._func_dict:
            now = time.time()
            # If cache is expired or missing, re-evaluate
            if key not in self._cache or (now - self._cache[key]["ts"]) > self._ttl:
                evaluated_val = self._func_dict[key](val, dyn_dict=self, key=key)
                self._cache[key] = {"val": evaluated_val, "ts": now}
            
            return self._cache[key]["val"]
        return val

    def __delitem__(self, key):
        if key in self._func_dict:
            del self._func_dict[key]
        if key in self._cache:
            del self._cache[key]
        super().__delitem__(key)
```

A Better Architectural Approach
Instead of a hardcoded time-based TTL, a cleaner, more reliable approach for Salt config structures is to use a cached property that flushes when the loader reloads, or a slightly longer, configurable default window (like 5 to 10 seconds).

If you want to keep the time-based approach but make it resilient against mid-job expiration, you can expose the TTL as a master configuration option that defaults to 5.0 seconds:

```python
self._ttl = opts.get("dynamic_dict_ttl", 5.0)
```

This gives you a safer buffer for long-running compilations while allowing teams with massive environments to turn it up if their disk I/O becomes a bottleneck.

twangboy commented on salt/utils/dynamic_dict.py:
```diff
@@ -0,0 +1,92 @@
+            val = default
+        return val
+
+    def values(self):
+        keys = super().keys()
+        for key in keys:
+            yield self[key]
```
Critical API Mismatch: ``.items()`` Omission and Broken Dictionary Views

The Issue: The DynamicDict implementation completely fails to override ``.items()`, and it implements ``.values()`` as a custom generator rather than a standard Python 3 dictionary view object (dict_values).

The Impact: Because ``.items()`` falls back to ``super().items()``, running loops like ``for env, paths in opts["file_roots"].items():`` bypasses the dynamic wrapper entirely, returning the raw, unglobbed static strings instead of the expanded paths. Furthermore, calling ``list(sdb_opts.items())`` in functions like ``apply_sdb`` fails to evaluate the dynamic keys, while code expecting standard dictionary view behaviors (such as set operations) will break when interacting with a raw generator.

The Fix: Explicitly override ``keys()``, ``values()``, and ``items()`` inside ``DynamicDict`` to evaluate the dynamic pathways cleanly and ensure parity with standard Python 3 dictionary behaviors.

Add the following methods to the ``DynamicDict`` class in ``salt/utils/dynamic_dict.py``, replacing the existing ``values()`` generator:

```python
    def keys(self):
        return super().keys()

    def values(self):
        # Return a custom list-like view or tuple list to mirror dict_values
        return [self[key] for key in super().keys()]

    def items(self):
        # Crucial fix: Ensures loops over .items() evaluate the dynamic paths
        return [(key, self[key]) for key in super().keys()]
```

twangboy commented on salt/utils/dynamic_dict.py:
```diff
@@ -0,0 +1,92 @@
+                rdd.add_dyn(key, func, data)
+            else:
+                rdd[key] = value
+            rdd[copy.deepcopy(key, memo)] = copy.deepcopy(value, memo)
```
Redundant Assignment Bug in ``__deepcopy__``
The Issue: Inside the else block of the ``__deepcopy__`` method, the implementation executes ``rdd[key] = value``. This assigns the uncopied references directly to the new dictionary, only for the very next line to completely overwrite it with a proper ``copy.deepcopy()`` operation.

The Impact: The initial assignment line is entirely redundant, adds unnecessary dict-write overhead during deep copy operations, and leaves confusing dead code in a core utility module.

The Fix: Clean up the ``else`` block by removing the redundant, non-copied assignment line entirely, ensuring the method handles copying cleanly.

Refactor the ``__deepcopy__`` loop in ``salt/utils/dynamic_dict.py`` like this:

```python
    def __deepcopy__(self, memo):
        rdd = DynamicDict()
        memo[id(self)] = rdd
        
        # Walk the underlying raw dictionary keys directly
        for key in super().keys():
            if key in self._func_dict:
                func = self._func_dict[key]
                data = super().__getitem__(key)
                rdd.add_dyn(key, func, data)
            else:
                # Fix: Strip out the redundant non-copied assignment line
                copied_key = copy.deepcopy(key, memo)
                copied_value = copy.deepcopy(super().__getitem__(key), memo)
                rdd[copied_key] = copied_value
                
        return rdd
```

Commit:
Address review feedback on DynamicDict: fix items()/deepcopy, add TTL cache

Fixes 3 issues from PR review: items() wasn't overridden so callers
(e.g. salt/client/ssh) iterating file_roots.items() got raw unglobbed
paths; __deepcopy__ had a redundant overwrite that becomes a real
double-evaluation bug once items() is fixed; and glob expansion ran
unbounded on every access. Adds a per-entry TTL cache (default 5s),
exposed as the dynamic_roots_ttl config option for file_roots,
pillar_roots, and thorium_roots.

Commit:
Fix import order in yamldumper.py (isort)
